File: 169-MajorityElement/169-MajorityElement.cpp
// Last updated: 16/10/2025, 04:28:35
class Solution {
public:
    int majorityElement(vector<int>& nums) {
        //  Boyer-Moore Voting Algorithm
        int res = 0;
        int count = 0;
        for(auto it : nums){
            if(count == 0){
                res = it;
                count++;
            }
            else if(it == res){
                count++;
            }
            else{
                count--;
            }
        }
        int cnt = 0;
        for(auto it : nums){
            if(it == res){
                cnt++;
            }
        }
        if(cnt > nums.size()/2){
            return res;
        }
        return -1;
    }
};

File: 169-MajorityElement/169-MajorityElement.java
// Last updated: 16/10/2025, 04:28:28
class Solution {
    //Boyer–Moore majority vote algorithm
    public int majorityElement(int[] nums) {
      int n = nums.length;
      int count = 0;
      int majorityElem = 0;

      for(int i = 0; i < n; i++){
        if(count == 0){
            majorityElem = nums[i];
            count++;
        }
        else if(nums[i] != majorityElem){
            count--;
        }else{
            count++;
        }
      }
      return majorityElem;
    }
}

